IPNI_FEDERATION.md

```diff
@@ -0,0 +1,406 @@
+  nodes, especially in a globally distributed setup, is not the objective.
+* **Network Topology Design**: The IPNI federation protocol does not define or restrict the underlying network topology.
+  Whether nodes are structured in a mesh, star, or hierarchical format, the protocol remains agnostic.
+* **Strong consistency**: While the IPNI federation protocol aims to ensure data consistency and availability, it does
+  not aim to atomically replicate every piece of indexed data across all nodes. For example, it is possible for some
+  records to remain partially available on one node until deletions are propagate through the network.
+
+## Terminology
+
+* **Epoch**: An epoch in the context of the IPNI federation protocol refers to a specific time period or iteration of
+  the protocol's operation. It can be thought of as a version or generation number that increments each time there's a
+  major operation or update, aiding in synchronization efforts across instances.
+
+* **Reconciliation**: Reconciliation is the process of ensuring that two or more IPNI instances have consistent index
+  records. It involves comparing the content of instances, identifying discrepancies, and making updates so that all
+  participating instances have the same content.
+
+* **Vector Clock (VC)**: Vector Clock is a mechanism used to determine the order of events in a distributed system. In
+  IPNI federation protocol, a vector clock helps track the sequence of index updates across multiple instances, thereby
+  assisting in maintaining the eventual consistency of the index.
+
+* **Snapshot**: A snapshot refers to a captured state of the index records at a particular point in time consisting of
+  the providers list known by the indexer, the latest processed advertisement CID for each provider and an epoch number.
+  IPNI instances can share snapshots with each other to quickly update and reconcile their content.
+
+## Specification
+
+The IPNI federation protocol is designed to address the challenges that arise when multiple IPNI instances operate
+concurrently across different regions or domains. The fundamental premise of this specification is to provide a
+systematic method for these instances to interact, synchronize, and offer a unified view of the index records to
+end-users. To achieve this, each node in a federated IPNI network performs the following:
+
+1. Initialization - the process by which a node joins a federated network.
+2. Periodic snapshots - the process by which a node periodically captures an immutable snapshot of its providers along
+   with associated metadata.
+3. Snapshot exchange - the process by which each node propagates their latest snapshot to the rest of the network.
+4. Reconciliation - the process by which snapshots received from other nodes is reflected on to bring the network state
+   knowledge up-to-date.
+
+Finally, the specification expands on cases where conflict might occur and how a node should resolve it.
+
```

Comment:
Should this specification propose ways to offer indexer incentives? Pay for querys, pay for results, pay for advertisements ingested, etc.?

Reply:
Answered above - No, that should be called out in a seperate spec as long as it is not precluded by this one.

Reply:
I think incentives is also a separate problem than "Can I find what I am looking for no matter which indexer I search?"

Those seem like time consuming factors to figure out which will most likely be application/service specific. I am not sure if we can meaningfully pin them down in a spec.

Another thing I would keep in mind is a wider ecosystem incentives for CIDs and content addressed data specifically. For example, how Boost works in terms of deal making or retrieval will impact incentives of indexing.

Overall, I too think it is something to certainly revisit. I would probably start with a concrete use case like what would those numbers be for cid.contact first before adding them to a spec.